File: algorithms/easy/69_sqrt_x.js
/**
 * 69. Sqrt(x)
 * 
 * @param {number} x
 * @return {number}
 */
var mySqrt = function(x) {
    // Handle edge cases
    if (x === 0 || x === 1) return x;
    
    // Binary search for the square root
    let left = 1;
    let right = x;
    
    while (left <= right) {
        const mid = Math.floor(left + (right - left) / 2);
        const square = mid * mid;
        
        if (square === x) {
            return mid;
        } else if (square < x) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    // When the loop ends, 'right' will be the integer square root of x
    return right;
};

// Alternative implementation using Newton's method
var mySqrtNewton = function(x) {
    if (x === 0 || x === 1) return x;
    
    let r = x;
    while (r > x / r) {
        r = Math.floor((r + x / r) / 2);
    }
    
    return Math.floor(r);
};

// Test cases
console.log(mySqrt(4));     // 2
console.log(mySqrt(8));     // 2
console.log(mySqrt(9));     // 3
console.log(mySqrt(16));    // 4
console.log(mySqrt(17));    // 4

File: algorithms/easy/69_sqrt_x.py
# 69. Sqrt(x)

def my_sqrt(x):
    """Compute the square root of x, truncated to the nearest integer.
    
    Args:
        x: Non-negative integer
    
    Returns:
        Integer square root of x, truncated to the nearest integer
    """
    # Handle edge cases
    if x == 0 or x == 1:
        return x
    
    # Binary search for the square root
    left, right = 1, x
    while left <= right:
        mid = left + (right - left) // 2
        square = mid * mid
        
        if square == x:
            return mid
        elif square < x:
            left = mid + 1
        else:
            right = mid - 1
    
    # When the loop ends, 'right' will be the integer square root of x
    return right

# Alternative implementation using Newton's method
def my_sqrt_newton(x):
    if x == 0 or x == 1:
        return x
    
    r = x
    while r > x / r:
        r = (r + x / r) // 2
    
    return int(r)

# Test cases
print(my_sqrt(4))     # 2
print(my_sqrt(8))     # 2
print(my_sqrt(9))     # 3
print(my_sqrt(16))    # 4
print(my_sqrt(17))    # 4